### DISCUSSION TITLE
Any easy way to implement unoccs on a monorepo with vite apps?

### DISCUSSION THREAD

**Discussion by velvetcode206** [unocss/unocss#915] [Q&A]
Hi guys! I have a `unocss.config.ts` file on the project root, where I define all themes, presets, colors, etc.

```js
import { defineConfig } from 'unocss';
import presetWind from '@unocss/preset-wind';
import presetWebFonts from '@unocss/preset-web-fonts';
import presetIcons from '@unocss/preset-icons';
import transformerDirective from '@unocss/transformer-directives';

/**
 * For more configuration options: https://github.com/unocss/unocss
 */
export default defineConfig({
  presets: [
    presetWind(),
    presetWebFonts({
      provider: 'google',
      fonts: {
       // My fonts.
      },
    }),
    presetIcons({
      extraProperties: {
        display: 'inline-block',
        'vertical-align': 'middle',
      },
    }),
  ],
  transformers: [transformerDirective()],
  theme: {
    // My theme property.
  },
  shortcuts: {
    // My shortcuts.
  },
});
```

My problem is when I'm importing it to a vite app, I presume this would be the correct way to do it (similar to tailwind and windi):

```js
import { defineConfig } from 'vite';
import unocss from 'unocss/vite';
import unocssConfig from '../../unocss.config';

export default defineConfig({
  plugins: [
    unocss(unocssConfig),
  ]
})
```

But it does't work, it seem the typing is wrong, `UserConfig<{all-those-parameters}>` is not assignable to parameter of type `'string | VitePluginConfig<{}> | undefined'`.

For now I got around this problem dissecting my config object like this:

```js
import { defineConfig } from 'vite';
import unocss from 'unocss/vite';
import unocssConfig from '../../unocss.config';

export default defineConfig({
  plugins: [
    unocss({
      presets: unocssConfig.presets,
      transformers: unocssConfig.transformers,
      theme: unocssConfig.theme,
      shortcuts: unocssConfig.shortcuts,
    }),
  ]
})
```

Is there a better way to do this kind of import without going over the parameter individually?

Backlog of the whole error: 
`Argument of type 'UserConfig<{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }>' is not assignable to parameter of type 'string | VitePluginConfig<{}> | undefined'.`

`Type 'UserConfig<{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }>' is not assignable to type 'VitePluginConfig<{}>'.`

`Types of property 'extendTheme' are incompatible.`

`Type 'Arrayable<ThemeExtender<{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }>> | undefined' is not assignable to type 'Arrayable<ThemeExtender<{}>> | undefined'.`

`Type 'ThemeExtender<{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }>' is not assignable to type 'Arrayable<ThemeExtender<{}>> | undefined'.`

`Type 'ThemeExtender<{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }>' is not assignable to type 'ThemeExtender<{}>'.`

`Type '{}' is missing the following properties from type '{ colors: { black: string; blue: { dark: string; light: string; }; gray: { dark: string; light: string; background: string; }; green: { dark: string; light: string; background: string; }; red: { dark: string; light: string; }; }; breakpoints: { ...; }; }': colors, breakpointsts(2345)`

**Comment by Cedric-ruiu** [comment]
I have the same errors. Does anyone know how to solve this problem?

**Comment by stafyniaksacha** [comment]
Maybe using `as const` solve it?

```ts
// unocss.config.ts
export default defineConfig({
  // ...
} as const);
```

**Comment by Cedric-ruiu** [comment]
@stafyniaksacha not really but thanks for helping.

```ts
// uno.config.ts
export default defineConfig({
  presets: [presetUno()],
  rules: [['text-xxs', { 'font-size': '0.625rem' }]],
  transformers: [transformerDirective()],
  include: ['./**/*.vue', /.*\/ui\/.*\.mjs(.*)?$/],
} as const);
```

```
Argument of type '{ readonly presets: readonly [Preset<Theme>]; readonly rules: readonly [readonly ["text-xxs", { readonly 'font-size': "0.625rem"; }]]; readonly transformers: readonly [SourceCodeTransformer]; readonly include: readonly [...]; }' is not assignable to parameter of type 'UserConfig<{}>'.
  Types of property 'rules' are incompatible.
    The type 'readonly [readonly ["text-xxs", { readonly 'font-size': "0.625rem"; }]]' is 'readonly' and cannot be assigned to the mutable type 'Rule<{}>[]'.
```

**Comment by zyyv** [comment]
Hi, we are supported input a config or a path at the same time.
```ts
export default defineConfig({
  plugins: [
    UnoCSS('../unocss.config.ts'),
  ]
})
```
https://github.com/unocss/unocss/blob/8554bf483da79bd51383a6f87f18160f90386a7d/packages/vite/src/index.ts#L27-L30

**Comment by Cedric-ruiu** [comment]
Thx @zyyv , but I have some trouble with string path parameter, it seems unrecognized.

I found a (ugly?) tricks to work in my monorepo project :

```ts
// vite.config.ts
import type { VitePluginConfig } from '@unocss/vite';
import UnoCSSConfig from '@my-monorepo/config-unocss/uno.config';

export default defineConfig({
  plugins: [
    UnoCSS(UnoCSSConfig as VitePluginConfig),
  ]
})
```

```ts
// uno.config.ts
import presetUno from '@unocss/preset-uno';
import transformerDirective from '@unocss/transformer-directives';
import type { VitePluginConfig } from '@unocss/vite';
import { defineConfig } from 'unocss';

const config: VitePluginConfig = defineConfig({
  presets: [presetUno()],
  rules: [['text-xxs', { 'font-size': '0.625rem' }]],
  transformers: [transformerDirective()],
});

export default config;
```
